pkg/driver/controller.go

```diff
@@ -101,109 +101,127 @@ func getRoundedCapacity(size int64) int64 {
 	return ((size + Mi - 1) / Mi) * Mi
 }
 
-func waitForReadyVolume(volname string) error {
-	for true {
-		vol, err := lvm.GetLVMVolume(volname)
-		if err != nil {
-			return status.Errorf(codes.Internal,
-				"lvm: wait failed, not able to get the volume %s %s", volname, err.Error())
+// waitForLVMVolume waits for completion of any processing of lvm volume.
+// It returns the final status of lvm volume along with a boolean denoting
+// whether it should be rescheduled on some other volume group or node.
+// In case volume ends up in failed state and rescheduling is required,
+// func is also deleting the lvm volume resource, so that it can be
+// re provisioned on some other node.
+func waitForLVMVolume(ctx context.Context,
+	vol *lvmapi.LVMVolume) (*lvmapi.LVMVolume, bool, error) {
+	var reschedule bool // tracks if rescheduling is required or not.
+	var err error
+	if vol.Status.State == lvm.LVMStatusPending {
+		if vol, err = lvm.WaitForLVMVolumeProcessed(ctx, vol.GetName()); err != nil {
+			return nil, false, err
 		}
+	}
+	// if lvm volume is ready, return the provisioned node.
+	if vol.Status.State == lvm.LVMStatusReady {
+		return vol, false, nil
+	}
 
-		switch vol.Status.State {
-		case lvm.LVMStatusReady:
-			return nil
+	// Now it must be in failed state if not above. See if we need
+	// to reschedule the lvm volume.
+	var errMsg string
+	if volErr := vol.Status.Error; volErr != nil {
+		errMsg = volErr.Message
+		if volErr.Code == lvmapi.InsufficientCapacity {
+			reschedule = true
```

Comment:
can we make is reschedule everytime there is failure on the  node, does not matter for what reason. If it fails on a perticular node, we would want to try the volume creation on some other node.

Reply:
I think it may not be a good idea to reschedule volume in case of any (possibly intermittent) error. It'll result in sparse & unpredictable scheduling of storage resources. IMO, if there is any error related to a particular node, it's better to inspect & fix that error(by cordoning the node) rather than masking it with rescheduling.